src/blob_file_builder.cc

```diff
@@ -6,18 +6,38 @@ namespace titandb {
 BlobFileBuilder::BlobFileBuilder(const TitanDBOptions& db_options,
                                  const TitanCFOptions& cf_options,
                                  WritableFileWriter* file)
-    : cf_options_(cf_options),
+    : builder_state_(cf_options.blob_file_compression_options.max_dict_bytes > 0
+                         ? BuilderState::kBuffered
+                         : BuilderState::kUnbuffered),
+      cf_options_(cf_options),
       file_(file),
-      encoder_(cf_options_.blob_file_compression) {
+      encoder_(cf_options_.blob_file_compression,
+               cf_options.blob_file_compression_options) {
   BlobFileHeader header;
+  if (cf_options.blob_file_compression_options.max_dict_bytes > 0)
+    header.flags |= BlobFileHeader::kHasUncompressionDictionary;
   std::string buffer;
   header.EncodeTo(&buffer);
   status_ = file_->Append(buffer);
 }
 
 void BlobFileBuilder::Add(const BlobRecord& record, BlobHandle* handle) {
   if (!ok()) return;
+  if (builder_state_ == BuilderState::kBuffered) {
+    std::string rec_str;
+    record.EncodeTo(&rec_str);
+    sample_records_.emplace_back(rec_str);
+    sample_str_len_ += rec_str.size();
+    if (cf_options_.blob_file_compression_options.zstd_max_train_bytes > 0 &&
+        sample_str_len_ >=
+            cf_options_.blob_file_compression_options.zstd_max_train_bytes) {
+      EnterUnbuffered();
+    }
+    return;
```

Comment:
`BlobGCJob::DoRun` will call `BlobFileBuilder::Add` and use the `BlobHandle` passed in. You should handle it correctly when in buffered state.

Reply:
OK, I will take a look at this part, I'm still new to titan, and definitely missed this detail. thanks for your reminder, will fix this later!

Reply:
I took a look roughly. The value of `handle` was not set when `Add` was called in the `buffered` state.
And, the offset and size of the `handle` can not be calculated in advance.
So currently, my idea is to use a callback-style design.
1. `BlobFileBuilder` holds a callback queue, when calling `Add` in buffered state, add an callable object to the callback queue, then call it(with related handle value) when the corresponding record is actually compressed and written.
2. When calling `Add` from `BlobGCJob` and `TitanTableBuilder`, check the state of the builder, if it is in buffered state, using callback to do the rest of work.


Is this design reasonable? Do you have any suggestions or points that I should be attention? Thanks.

Reply:
I think we can buffer the key first, and returned the blob handle by `BlobFileBuidler->Finish` and write the key and handle at that time

Reply:
Also, please note `blob_builder_->Add` in `TitanTableBuilder::AddBlob` suffers from the same issue. Hmmmm, seems the code changes would be far more than I expected.

Reply:
Instead of callback, how about just make caller of `Add()` pass in a `vector<BlobHandle*>`, and batch return the handles when the builder exit buffer state?

Better still, instead of vector you can use `autovector` in rocksdb code, which optimize for small vector.

Reply:
@yiwu-arbug sounds much simpler, I will try this way

Reply:
> I think we can buffer the key first, and return the blob handle by `BlobFileBuidler->Finish` then write the key and handle at that time. What do you think @yiwu-arbug

ah, didn't see this comment. I think this is a better idea. @chux0519

Reply:
I guess add another param(vector<BlobHandle*>) would be easier. In this way, we only need to check if the vector is empty or not before using handles we cant. Move the write logic to `Finish` sounds like I will refactor the caller function logic.
Still not sure, I will look into related code closer later.

Reply:
You do need to refactor the caller, because there's still chance `EnterUnbuffer()` is done on `Finish()`.